### DISCUSSION TITLE
I am having difficulties coloring the tracts with FA maps

### DISCUSSION THREAD

**Discussion by drsiyarb** [dipy/dipy#2712] [Scientific Visualization]
Hello everyone, 

I am currently working on a dataset from my organization.

I needed to visualize a 3D image of ventricles and some segments of corticospinal tracts for anatomical visualization. I have prepared .trk files with dipy before, using `CsaOdfModel` and `LocalTracking`. 

I was able to follow advanced viz tutorial, and generated very cool images with directionality information, but for the purposes of my research,I wanted to see the tractograms colored with FA maps therefore I started to follow this tutorial; 

https://dipy.org/documentation/1.5.0/examples_built/viz_bundles/#example-viz-bundles

Unfortunately, I didnt prepare FA maps beforehand because I had the idea later. So I tried to calculate an FA map from one pateint and try bulk processing later. Here is the function I use;

```
import nibabel as nib
from dipy.reconst.dti import TensorModel
from dipy.reconst.dti import fractional_anisotropy, color_fa
from dipy.core.gradients import gradient_table



def extract_dti(dwi_file, bvals, bvecs):
    # Load the DWI data
    img = nib.load(dwi_file)
    data = img.get_fdata()

    # Get the affine of the image
    affine = img.affine

    # Get the header of the image
    header = img.header
    
    # Create a gradient table
    gtab = gradient_table(bvals, bvecs)
    # Create the tensor model
    model = TensorModel(gtab)

    # Fit the data to the tensor model
    tensor_fit = model.fit(data)
    
    # Compute the FA
    FA = fractional_anisotropy(tensor_fit.evals)

    FA = np.clip(FA, 0, 1)
    RGB = color_fa(FA, tensor_fit.evecs)
    nib.save(nib.Nifti1Image(np.array(255 * RGB, 'uint8'), img.affine), 'tensor_rgb.nii.gz')

    return FA, RGB, affine, header
```

I try to use the code below, the 3D tract is rendered with a solid color... No sign of FA map.  

```
scene = window.Scene()


hue = (0.0, 0.0)  # red only
saturation = (0.0, 1.0)  # white to red

lut_cmap = actor.colormap_lookup_table(hue_range=hue,
                                       saturation_range=saturation)

stream_actor = actor.line(streamlines, linewidth=0.1, colors=FA)
bar2 = actor.scalar_bar()

if not world_coords:
    image_actor_z = actor.slicer(data, affine=np.eye(4))
else:
    image_actor_z = actor.slicer(data, affine)
```

 What might my problem be?

**Comment by drsiyarb** [comment]
![fa_bundle](https://user-images.githubusercontent.com/16266287/212014803-2c7a7285-8510-4bf4-8775-405c6679a909.jpg)

This is how my bundle looks. 

When I run the tutorial data, it runs perfectly and produces these images;
![bundle3](https://user-images.githubusercontent.com/16266287/212014914-f6a5d86c-fc21-41d5-842e-56c7c665d25f.jpg)
![bundle1](https://user-images.githubusercontent.com/16266287/212014918-a708a855-eed5-428a-bd69-b69db6ab764c.jpg)

**Comment by skoudoro** [comment]
Hi @drsiyarb,

Not sure but it looks like a color range issue. Your streamlines are purple which means close to zero following your scalar bar. 

Can you precise it on your colormap? tell us if it helps. 

```python
lut_cmap = actor.colormap_lookup_table(scale_range=(fa.min(), fa.max()*0.8), hue_range=hue, saturation_range=saturation, value_range=(0., 1.))
```

**Reply by drsiyarb** [reply]
Thank you very much @skoudoro, I went on to plot a histogram of the image values in my FA map, you can see them below. Your suspicions about the color range were true, there is an incredible amount of voxels at the both 0 and 1 so I tried to adjust the scale range in accordance with histogram , then when I adjust the scale range between 0.05 and 0.8 but unfortunately, I still get the same purple output, no change at all.

Code for histogram:

```
import nibabel as nib
import matplotlib.pyplot as plt


data=FA
# Flatten the 3D data into a 1D array
data = data.flatten()

# Plot the histogram
plt.hist(data, bins=250)
plt.xlabel("Value")
plt.ylabel("Frequency")
plt.title("Histogram of Image Values")
plt.show()
```

The histogram:

**Reply by drsiyarb** [reply]
![output](https://user-images.githubusercontent.com/16266287/212464900-01b56cbd-eab4-47fb-8061-e6f3df5ae64d.jpg)
